lib/ansible/module_utils/api.py

```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
 
         return retried
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay=3, max_delay=60):
+    """The "Full Jitter" backoff strategy described here https://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    :param retries: The number of delays to generate.
+    :param delay: The time in seconds for the initial delay.
+    :param max_delay: The maximum time in seconds for any delay.
+    """
+    for retry in range(0, retries):
+        yield random.randint(0, min(max_delay, delay * 2 ** retry))
+
+
+def retry_with_delays_and_condition(backoff_iterator, retry_condition, do_on_retry=None, retry_on_success=None):
+    """Generic retry decorator.
+
+    :param backoff_iterator: An iterable of delays in seconds.
+    :param retry_condition: A callable that takes an exception of the decorated function and decides whether to retry or not (returns a bool).
+    :param do_on_retry: An optional callable that takes an exception/None and the current delay and runs before the sleep (returns None).
+    :param retry_on_success: An optional callable that accepts the decorated function return value and decides whether or not to retry (returns a bool).
+    """
+    def function_wrapper(function):
+        def run_function(*args, **kwargs):
+            """This assumes the function has not already been called.
+            If backoff_iterator is empty, we should still run the function a single time with no delay.
+            """
+
+            for delay in backoff_iterator:
+                error = None
+
+                try:
+                    result = function(*args, **kwargs)
+                except Exception as e:
+                    if not retry_condition(e):
+                        raise
+                    error = e
+                else:
+                    if not callable(retry_on_success) or not retry_on_success(result):
+                        return result
+
+                if callable(do_on_retry):
+                    do_on_retry(error, delay)
```

Comment:
Same here:
```suggestion
                do_on_retry(error, delay)
```
and
```python
if do_on_retry is None:
    do_on_retry = lambda *_, *__: None
```
outside `function_wrapper`